nni/mutable/mutable.py

```diff
@@ -0,0 +1,482 @@
+
+        Parameters
+        ----------
+        memo
+            A dict of mutable labels and their random values.
+            Use this to share the sampled value among mutables with the same label.
+        """
+        sample: Sample = {} if memo is None else memo
+        if random_state is None:
+            random_state = RandomState()
+        for mutable in self.simplify().values():
+            mutable.random(sample, random_state)
+        return self.freeze(sample)
+
+    def grid(self, memo: Sample | None = None, granularity: int | None = None) -> Iterable[Any]:
+        """Return a grid of sample points
+        that can be possibly sampled from the mutable. Used in grid search strategy.
+        It should return all the possible results of :meth:`freeze`.
+
+        The default implementation of :meth:`grid` is to call iterate over
+        the product of all the simplified grid values.
+        Specifically, the grid will be iterated over in a depth-first-search order.
+
+        Parameters
+        ----------
+        memo
+            A dict of mutable labels and their values in the current grid point.
+            Use this to share the sampled value among mutables with the same label.
+        granularity
+            Optional integer to specify the level of granularity of the grid.
+            This only affects the cases where the grid is not a finite set.
+            See :class:`Continuous` for details.
+        """
+        def _iter(index: int) -> Iterable[Any]:
+            if index == len(simplified):
+                yield self.freeze(sample)
+            else:
+                for _ in simplified[index].grid(sample, granularity):
+                    yield from _iter(index + 1)
+
+        # No deduplication here as it will be done in the grid of simplified mutables.
+        simplified: list[LabeledMutable] = list(self.simplify().values())
+
+        # Same sample is used throughout the whole process.
+        sample: Sample = {} if memo is None else memo
+
+        yield from _iter(0)
+
+
+class LabeledMutable(Mutable):
+    """:class:`Mutable` with a label. This should be the super-class of most mutables.
+    The labels are widely used in simplified result, as well as samples.
+    Usually a mutable must be firstly converted into one or several :class:`LabeledMutable`,
+    before strategy can recognize and process it.
+
+    When two mutables have the same label, they semantically share the same choice.
+    That means, the choices of the two mutables will be shared.
+
+    The labeled mutables are by default created with one reproducible, auto-generated label.
+    But it can also contain multiple labels if the mutable is more complex.
+    The label can also be specified by users, or sub-classes.
+
+    Being a :class:`LabeledMutable` doesn't necessarily mean that it is a leaf mutable.
+    Some :class:`LabeledMutable` can be further simplified into multiple leaf mutables.
+    """
+
+    label: str
```

Comment:
Isn't this static?

Reply:
It's a instance variable. This is [Python syntax](https://peps.python.org/pep-0526/#class-and-instance-variable-annotations).

Reply:
Wow, I though it was just type hint.